### PR TITLE
Migrate investments to angular

### PR DESCRIPTION
**Pull request by AlexPoliteRudenko**

### CODE REVIEW

File: src/main/resources/static/views/investment/Investments.html [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                    </div>
+                    <div class="col-xs-2">
+                        {{investment.sum | currency}}
+                    </div>
+                </div>
+            </li>
+        </ul>
+    </div>
+</div>
+<!--
```
Comment (Sheptytskyid) [inline src/main/resources/static/views/investment/Investments.html]: MAJOR Remove this commented out code. [rule](https://sonarqube.com/coding_rules#rule_key=Web%3AAvoidCommentedOutCodeCheck)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/main/java/net/greatstart/mappers/InvestmentMapper.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package net.greatstart.mappers;
+
+import net.greatstart.dto.DtoInvestment;
+import net.greatstart.model.Investment;
+import org.mapstruct.Context;
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+@Mapper(componentModel = "spring")
+public interface InvestmentMapper {
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/mappers/InvestmentMapper.java]: CRITICAL Move constants to a class or enum. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1214)

File: src/main/java/net/greatstart/dto/DtoProject.java [hunk 3]
```diff
@@ -14,4 +15,6 @@
     @Valid
     private DtoProjectDescription desc;
 
+    private List<DtoInvestment> dtoInvestments;
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/dto/DtoProject.java]: MAJOR Remove this unused "dtoInvestments" private field. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1068)

File: src/main/java/net/greatstart/mappers/ProjectMapper.java [hunk 4]
```diff
@@ -1,12 +1,28 @@
...
 import net.greatstart.model.Project;
+import org.mapstruct.Context;
 import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.factory.Mappers;
+
+import java.util.List;
 
 @Mapper(componentModel = "spring")
 public interface ProjectMapper {
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/mappers/ProjectMapper.java]: CRITICAL Move constants to a class or enum. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1214)

File: src/main/resources/static/views/investment/ProjectInvestments.html [hunk 5]
```diff
@@ -0,0 +1,31 @@
+<div ng-controller="ProjectController"
+     style="max-height: 210px; overflow: auto;">
+	<ul ng-repeat="investment in project.dtoInvestments"
+	    class="list-group checked-list-box">
+		<!--ng-if="project.investor.email == currentUser."-->
```
Comment (Sheptytskyid) [inline src/main/resources/static/views/investment/ProjectInvestments.html]: MAJOR Remove this commented out code. [rule](https://sonarqube.com/coding_rules#rule_key=Web%3AAvoidCommentedOutCodeCheck)

File: src/main/java/net/greatstart/controllers/ProjectController.java [hunk 6]
```diff
@@ -72,7 +76,7 @@ public ModelAndView showMyProjects(Principal principal) {
         return model;
     }
 
-    @GetMapping(value = "/{id}")
+    /*@GetMapping(value = "/{id}")
     public ModelAndView showProject(@PathVariable Long id) {
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/controllers/ProjectController.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/main/java/net/greatstart/dto/DtoProject.java [hunk 7]
```diff
@@ -13,5 +14,8 @@
     private DtoCategory category;
     @Valid
     private DtoProjectDescription desc;
+    private DtoUserProfile owner;
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/dto/DtoProject.java]: MAJOR Remove this unused "owner" private field. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1068)

File: src/main/java/net/greatstart/dto/DtoProject.java [hunk 8]
```diff
@@ -13,5 +14,8 @@
...
+
+    private List<DtoInvestment> dtoInvestments;
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/dto/DtoProject.java]: MAJOR Remove this unused "dtoInvestments" private field. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1068)

File: src/main/resources/static/views/project/ProjectPage.html [hunk 9]
```diff
@@ -1,4 +1,117 @@
...
+					<div class="caption">
+						<h3>Other</h3>
+						<hr>
+						<p></p>
+					</div>
+				</div>
+			</div>
+			<!--Investments-->
+			<!--todo: hide investments if user is not project owner	-->
+			<!--<div ng-if="currentUser!=null && currentUser.email == project.owner.email">-->
```
Comment (Sheptytskyid) [inline src/main/resources/static/views/project/ProjectPage.html]: MAJOR Remove this commented out code. [rule](https://sonarqube.com/coding_rules#rule_key=Web%3AAvoidCommentedOutCodeCheck)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 4 major

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/main/java/net/greatstart/controllers/InvestmentController.java [hunk 10]
```diff
@@ -1,131 +1,72 @@
...
+        return new ResponseEntity<>(HttpStatus.EXPECTATION_FAILED);
     }
 
-    private BigDecimal getInvestedSumInProject(long id) {
-        return projectService.getProjectById(id).getInvestments()
-                .stream().map(Investment::getSum).reduce(BigDecimal.ZERO, BigDecimal::add);
+    @PutMapping("{id}")
+    public ResponseEntity<DtoInvestment> updateInvestment(@PathVariable long id,
+                                                          @RequestBody DtoInvestment investment) {
+        //todo update investment
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/controllers/InvestmentController.java]: INFO Complete the task associated to this TODO comment. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/net/greatstart/dto/DtoUserProfile.java [hunk 11]
```diff
@@ -17,6 +20,7 @@
     private String email;
     private String lastName;
     private byte[] photo;
+    private List<DtoInvestment> dtoInvestments;
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/dto/DtoUserProfile.java]: MAJOR Remove this unused "dtoInvestments" private field. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1068)

File: src/main/java/net/greatstart/configs/SpringSecurityDev.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
+import org.springframework.security.config.annotation.method.configuration.EnableGlobalMethodSecurity;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.csrf.CookieCsrfTokenRepository;
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/configs/SpringSecurityDev.java]: MINOR Remove this unused import 'org.springframework.security.web.csrf.CookieCsrfTokenRepository'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/net/greatstart/configs/SpringSecurityDev.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+                        "/views/main/Help.html",
+                        "/views/user/UserPage.html",
+                        "/views/user/EditUser.html").authenticated()
+                .and()
+                .authorizeRequests().antMatchers("/console/**").permitAll()
+                .and()
+                .formLogin()
+                .loginPage("/user/login")
+                .usernameParameter("email")
+                // TODO: change the routing when home page is ready
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/configs/SpringSecurityDev.java]: INFO Complete the task associated to this TODO comment. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/net/greatstart/controllers/InvestmentController.java [hunk 14]
```diff
@@ -1,131 +1,72 @@
...
+        return new ResponseEntity<>(HttpStatus.EXPECTATION_FAILED);
     }
 
-    private BigDecimal getInvestedSumInProject(long id) {
-        return projectService.getProjectById(id).getInvestments()
-                .stream().map(Investment::getSum).reduce(BigDecimal.ZERO, BigDecimal::add);
+    @PutMapping("{id}")
+    public ResponseEntity<DtoInvestment> updateInvestment(@PathVariable long id,
+                                                          @RequestBody DtoInvestment investment) {
+        //todo: implement update investment
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/controllers/InvestmentController.java:71]: INFO Complete the task associated to this TODO comment. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/net/greatstart/dto/DtoProjectDescription.java [hunk 15]
```diff
@@ -21,4 +22,53 @@
     private LocalDate dateStart;
     private LocalDate dateAdded;
     private byte[] image;
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/dto/DtoProjectDescription.java]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 4 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/net/greatstart/validators/InvestmentValidationService.java [hunk 16]
```diff
@@ -15,23 +17,30 @@
             + "Max possible investment is: $";
     private static final String WRONG_STEP_VALUE = MESSAGE
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/validators/InvestmentValidationService.java]: MAJOR Remove this unused "WRONG_STEP_VALUE" private field. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1068)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 7 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [InvestmentValidationService.java#L14](https://github.com/Sheptytskyid/Startup_Platform/blob/ed98a52645438c6b352a1d63c0ad9b6d53237bec/src/main/java/net/greatstart/validators/InvestmentValidationService.java#L14): Remove this unused "WRONG_MIN_VALUE" private field. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1068)
1. ![MAJOR][MAJOR] [InvestmentValidationService.java#L16](https://github.com/Sheptytskyid/Startup_Platform/blob/ed98a52645438c6b352a1d63c0ad9b6d53237bec/src/main/java/net/greatstart/validators/InvestmentValidationService.java#L16): Remove this unused "WRONG_MAX_VALUE" private field. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1068)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 4 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: src/main/java/net/greatstart/controllers/InvestmentController.java [hunk 17]
```diff
@@ -1,131 +1,83 @@
...
 
-
-    @GetMapping("/investment/{id}/delete")
-    public ModelAndView deleteInvestmentById(@PathVariable long id) {
-        investmentService.deleteInvestment(id);
-        return new ModelAndView(REDIRECT + INVESTMENTS_PAGE);
+    @PostMapping("")
+    public ResponseEntity<DtoInvestment> createInvestment(@Valid @RequestBody DtoInvestment investment) {
+        investment.setDateOfInvestment(LocalDateTime.now().truncatedTo(ChronoUnit.SECONDS));
+        investment.setVerified(false);
```
Comment (zozich) [inline src/main/java/net/greatstart/controllers/InvestmentController.java:57]: 1. Why duplicate the line?
2. Remove all business logic from controllers

**Review by zozich — APPROVED** [review]
Ok for now...

**Pull request merged**